x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -252,46 +292,128 @@ export class ExecutionHandler<
         }
       }
     }
+    return summaryActions;
   }
 
-  private generateExecutables(
-    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>
-  ) {
-    const executables = [];
+  private isAlertMuted(alertId: string) {
+    const muted = this.mutedAlertIdsSet.has(alertId);
+    if (muted) {
+      if (
+        !this.skippedAlerts[alertId] ||
+        (this.skippedAlerts[alertId] && this.skippedAlerts[alertId].reason !== Reasons.MUTED)
+      ) {
+        this.logger.debug(
+          `skipping scheduling of actions for '${alertId}' in rule ${this.ruleLabel}: rule is muted`
+        );
+      }
+      this.skippedAlerts[alertId] = { reason: Reasons.MUTED };
+      return true;
+    }
+    return false;
+  }
 
-    for (const action of this.rule.actions) {
-      for (const [alertId, alert] of Object.entries(alerts)) {
-        const actionGroup = this.getActionGroup(alert);
+  private isExecutableAction(action: RuleAction) {
+    return this.taskRunnerContext.actionsPlugin.isActionExecutable(action.id, action.actionTypeId, {
+      notifyUsage: true,
+    });
+  }
 
-        if (!this.ruleTypeActionGroups!.has(actionGroup)) {
-          this.logger.error(
-            `Invalid action group "${actionGroup}" for rule "${this.ruleType.id}".`
-          );
-          continue;
-        }
+  private isRecoveredAlert(actionGroup: string) {
+    return actionGroup === this.ruleType.recoveryActionGroup.id;
+  }
 
-        if (action.group === actionGroup && !this.isAlertMuted(alertId)) {
-          if (
-            this.isRecoveredAlert(action.group) ||
-            this.isExecutableActiveAlert({ alert, action })
-          ) {
-            executables.push({ action, alert });
-          }
+  private isSummaryAction(action: RuleAction) {
+    return action.frequency?.summary || false;
+  }
+
+  private isActionOnThrottleInterval(action: RuleAction) {
+    if (!action.frequency) {
+      return false;
+    }
+    return (
+      action.frequency.notifyWhen === RuleNotifyWhenTypeValues[2] &&
+      typeof action.frequency.throttle === 'string'
+    );
+  }
+
+  private isSummaryActionThrottled(action: RuleAction) {
```

Comment:
For some of these more complicated private functions, I would like to see them extracted into a library and unit tested.

Reply:
moving

Reply:
Done in [#3d137a1](https://github.com/elastic/kibana/pull/147360/commits/3d137a13893d66c7b29ac93e20446f4fefbba12c)